### PR TITLE
Practice dodge when damaged

### PR DESCRIPTION
**Pull request by Coolthulhu**
Many lines changed, because I implemented the change in `Character` rather than `player`. This required moving `practice` and thus `focus_pool` to `Character`.

Dodge is now trained when damaged by melee attack, projectile attack and impact damage (when landing after a smash).
Training is proportional to damage.

It's still not worth getting damaged to train dodge, but now moderately armored characters characters actually have a chance to get it up without explicit grinding. It also is a consolation prize to all those characters who haven't yet learned how to properly protect themselves.
It's tied to damage rather than just "getting hit" because otherwise it would be all about armoring self and getting pounded by weak stuff.

Lowered dodge training from boomer bile heavily. It is no longer possible to gear up in resistant stuff and engage in emetophilic fantasies to gain 10 dodge while risking only time and dignity. Getting boomer'd by a regular boomer will train it only up to 3, huge one up to 5.

### CODE REVIEW

**Comment by kevingranade** [comment]
This is nonsensical, especially scaling it with damage.
How I'd like it to work is for dodge to have a relatively easy to achieve
partial success that triggers practicing the skill.  I know overhauling how
dodge works is too much to ask for, but how about adding dodge practicing
to weapon blocking?  IIRC that has the partial success type mechanic in
place.

**Comment by Coolthulhu** [comment]
> This is nonsensical, especially scaling it with damage.

Scaling it with damage is correlated with need to actually get it right. It is the difference between someone in full plate dodging punches for fun and someone who actually needs it to survive.
It is no more nonsensical than only awarding practice for successfully dodging.

**Comment by kevingranade** [comment]
The rationale for success triggering practice is that since it worked, you
have positive feedback. I'd be OK with each attempt to dodge causing a tiny
amount of practice, but not scaling it with damage. In that case dodge
attempts should also have a cost (they might already, don't remember). I
consider finding a monster that can't or is unlikely to damage you and
using it to practice a feature, not a bug. To prevent it from making dodge
training trivial, the passive dodge practice could have a level cap, so you
could practice to lvl 1-3 or so with a mixture of successes and failures,
but after that you'd need to succeed to practice.

Some of the absurd scenarios damage-based skilling causes:
An apparently identical attack can either always or never train dodge
because of aspects of the attack the player is unaware of.
An identical attack will sometimes train dodging and sometimes not based on
the attack roll, which is unknown to both the player and the game until
after success or failure of dodging is determined.
A player who is so encumbered they can't move can still learn dodging.
A less armored character will learn more from the same attack, even at the
same encumbrance level.

**Comment by Coolthulhu** [comment]
> I consider finding a monster that can't or is unlikely to damage you and using it to practice a feature, not a bug.

That produces a very unnatural and anti-fun situation where training dodge naturally is a bad idea and grinding it is intentionally rewarded.

As dodges are probabilistic, it isn't necessary to cross any dodge value threshold to gain successes. So the only thing needed here is an armor that is good enough to guarantee safety from the training monster and light enough to allow dodging once per 40 tries or so.
Both are readily available, the only thing that may not be is player's willingness to put up with grinding and other bad game mechanics.

Note that the above "sanctioned exploit" doesn't have any real in-game cost. Just player cost, as the player needs to disengage once in a while to repair damaged clothing (cotton, so repairs are nearly free), making it a bit more tedious than just holding '5'.

At the same natural training, by going into combat and dodging attacks, is unrewarding, dangerous, slow and a bad idea.
The only real way to naturally train dodge is by getting heavily armored and then fighting lots of weak stuff with a weapon too weak to take them down quickly.

The problem is that to dodge (regular melee) attacks reliably, you need not to need to do so.
Like some weird Zen idea: to learn Dodging, unlearn Dodging (ie. learn to live as if you couldn't dodge anything at all).
Or grind it.
Just don't do anything in between those two.

Damage scaling isn't perfect, but our current dodge is just broken.

**Comment by ProfoundDarkness** [comment]
If I could chime in...  There are some situations where learning from failure is at least as... impactful, as learning from success.  Especially when pain is involved.

I don't know how much one could learn from being thrown (without armor) through a wall, besides "don't," but there is a lot of room between barely hurt and said wall where I'd expect some learning to happen.  Even that it's scaled based on damage makes some sense though I might expect a limit on how much more one can learn from larger doses of pain relative to others.

This next one is out there but I can see a little learning happening with no damage failed dodges though within that particular case I'd say it would be apt to cap that pretty low.  This really only makes much sense if there is indeed a cost to dodge attempts.

**Comment by Lady-Luck** [comment]
Just curious when this is going to make it in?

**Comment by Zireael07** [comment]
Seconding the question :)

**Comment by mugling** [comment]
This has a real potential to delay #18351 and the refactors moving code from `player` to `Character` are welcome from a code-sanity POV so I'd like them to go in if nothing else.

This PR isn't perfect but is a considerable improvement. Many of the issues raised do still need tackling but I think an approach of incremental improvement is reasonable especially given both @Coolthulhu's track record and continuing submission of PR's on the subject.

The only critical flaw I can see that needs solving now is this:

> A player who is so encumbered they can't move can still learn dodging.

Also needs rebase

**Comment by Coolthulhu** [comment]
I can rebase, but I don't have an obvious solution to the immobile player part.

File: src/character.cpp [hunk 1]
```diff
@@ -2052,3 +2055,145 @@ long Character::ammo_count_for( const item &gun )
...
+        practice( skill_dodge, 5 * total_damage, 10 );
+    }
+}
+
+void Character::deal_projectile_attack( Creature *source, dealt_projectile_attack &attack )
+{
+    Creature::deal_projectile_attack( source, attack );
+    const int total_damage = attack.dealt_dam.total_damage();
+    if( total_damage > 0 ) {
+        practice( skill_dodge, 5 * total_damage, 10 );
```
Comment (mugling) [inline src/character.cpp]: Could you not limit xp relative to encumbrance here?

Comment (Coolthulhu) [inline src/character.cpp]: Just the amount or level too?

Comment (mugling) [inline src/character.cpp]: Amount should suffice as a temporary measure until we get a better implementation of dodges/blocks

**Comment by Coolthulhu** [comment]
Training with damage now linearly decreases with sum of encumbrance to legs and torso.
Legs are worth half the torso (so you need 1 point on both legs to count as much as 1 on torso).
The scaling is `5.0 - ( (leg_l+leg_r)/2+torso ) / 10`. That is, someone with 50 encumbrance on torso won't get those on-damage xp gains, someone with 25 on each leg and 0 on torso will get 50% the gain (but 100% the focus drop), a nudist will get 100% the gains.

**Comment by mugling** [comment]
Trains way too quickly for damage - needs to be ~10% of current

**Comment by YeAdventurer** [comment]
Is this still happening?

**Comment by Lady-Luck** [comment]
Now that the new dodge formula is in, I assume this isn't far behind?

**Comment by Zireael07** [comment]
Why is it suddenly unmergeable?

**Comment by Lady-Luck** [comment]
>Why is it suddenly unmergeable?

It probably needs to be rebased.

**Comment by mugling** [comment]
Do you want to continue with this? Needs a rebase and XP gain scaling to 10% of original

**Comment by Coolthulhu** [comment]
At 10% of the original, it would be literally unnoticeable after level 2, not worth reviving.

**Comment by Zireael07** [comment]
What 'original' are you referring to? What was before the PR or what was PR'ed?

**Comment by mugling** [comment]
Can you otherwise adjust scaling - it's far too generous at low levels?

**Comment by Lady-Luck** [comment]
>At 10% of the original, it would be literally unnoticeable after level 2, not worth reviving.

Correct me if I'm wrong, but getting those first few levels of dodge would help significantly for training dodge further, ya?

Maybe change it to a curve?

**Comment by Lady-Luck** [comment]
Yes, no, maybe so?

**Comment by Coolthulhu** [comment]
I'm closing this one unless someone has a full idea on how to fix it.
Reducing to 10% will make it not noticeable above level 1.

New dodge formulas mean dodges against accurate enemies happen much more often, meaning it may no longer be necessary, though it could still be useful to prevent tedious dodge grind. Just could - I don't have a good use case here.

**Comment by Lady-Luck** [comment]
>I'm closing this one unless someone has a full idea on how to fix it.
Reducing to 10% will make it not noticeable above level 1.

Again, why not make it so getting hit becomes more effective for training dodge as your dodge levels up?

**Comment by Coolthulhu** [comment]
> Again, why not make it so getting hit becomes more effective for training dodge as your dodge levels up?

I don't have good numbers here.
The 100% version is barely noticeable at lvl 3-4, meaning we'd need something that would spike from 10% at 0 to 200% at lvl 5.

**Comment by Lady-Luck** [comment]
>The 100% version is barely noticeable at lvl 3-4, meaning we'd need something that would spike from 10% at 0 to 200% at lvl 5.

Does it necessary NEED to be? Just having this 10% thing in place would be good for characters just starting out, and then they could rely on skilling up through actually dodging.

**Comment by Lady-Luck** [comment]
@Coolthulhu Do you plan on picking this back up? New characters that have no ranks in dodge don't stand much of a chance of training it up without finding a book on dodging first - even if this becomes negligible after a few ranks it's something I think would be very appreciated.

**Comment by Coolthulhu** [comment]
I'd need to know how much practice is too much at which level.

**Comment by mugling** [comment]
> I'm closing this one unless someone has a full idea on how to fix it.

It doesn't look like this is the case at the moment but of course it could be reopened if anyone, especially @Coolthulhu, does work it out

**Pull request closed without merging**